aten/src/ATen/native/mps/operations/HistogramKernel.mm

```diff
@@ -0,0 +1,395 @@
+    T element = input_[offsets[tid * num_dims + dim]];
+
+    // Skips elements which fall outside the specified bins and NaN elements
+    // Adding an eps to the edges to eliminate precision issues that cause elements accidentally skipped,
+    // this is likely due to the minuscule implementation differences between the CPU and MPS's linspace.
+    if (!(element >= (leftmost_edge[dim] - eps) && element <= (rightmost_edge[dim] + eps))) {
+        skip_element = true;
+        break;
+    }
+    int64_t pos = -1;
+
+    if (algorithm == BIN_SELECTION_ALGORITHM::BINARY_SEARCH) {
+      pos = upper_bound(
+        bin_seq,
+        bin_seq_offset,
+        num_bin_edges[dim],
+        element
+      ) - bin_seq_offset - 1;
+    } else if (
+      algorithm == BIN_SELECTION_ALGORITHM::LINEAR_INTERPOLATION ||
+      algorithm == BIN_SELECTION_ALGORITHM::LINEAR_INTERPOLATION_WITH_LOCAL_SEARCH) {
+      pos = static_cast<int64_t>((element - leftmost_edge[dim])
+                            * (num_bin_edges[dim] - 1)
+                            / (rightmost_edge[dim] - leftmost_edge[dim]));
+      if (algorithm == LINEAR_INTERPOLATION_WITH_LOCAL_SEARCH) {
+          int64_t pos_min = max(static_cast<int64_t>(0), pos - 1);
+          int64_t pos_max = min(pos + 2, num_bin_edges[dim]);
+          pos = upper_bound(
+            bin_seq,
+            bin_seq_offset + pos_min,
+            pos_max - pos_min,
+            element
+          ) - bin_seq_offset - 1;
+      }
+    }
+
+    if (pos == (num_bin_edges[dim] - 1)) {
+      pos -= 1;
+    }
+    hist_index += local_out_strides[dim + 1] * pos;
+    bin_seq_offset += num_bin_edges[dim];
+  }
+  if (!skip_element) {
+    // In the unweighted case, the default weight is 1
+    local_out[local_out_strides[0] * tid + hist_index] += has_weight ? weight[tid] : 1;
+  }
+}
+
+
+#define REGISTER_HISTOGRAMDD_OP(DTYPE)                        \
+template                                                      \
+[[host_name("histogramdd_" #DTYPE)]]                          \
+kernel void histogramdd<DTYPE>(                               \
+  constant DTYPE    * input_                  [[buffer(0)]],  \
+  constant DTYPE    * weight                  [[buffer(1)]],  \
+  device   DTYPE    * local_out               [[buffer(2)]],  \
+  constant uint     * offsets                 [[buffer(3)]],  \
+  constant size_t   & num_dims                [[buffer(4)]],  \
+  constant DTYPE    * bin_seq                 [[buffer(5)]],  \
+  constant int64_t  * num_bin_edges           [[buffer(6)]],  \
+  constant DTYPE    * leftmost_edge           [[buffer(7)]],  \
+  constant DTYPE    * rightmost_edge          [[buffer(8)]],  \
+  constant int64_t  * local_out_strides       [[buffer(9)]],  \
+  constant uint8_t  & bin_selection_algorithm [[buffer(10)]], \
+  constant uint8_t  & has_weight              [[buffer(11)]], \
+  uint tid [[thread_position_in_grid]]);
+
+REGISTER_HISTOGRAMDD_OP(float);
+//REGISTER_HISTOGRAMDD_OP(half);
```

Comment:
Are we planning to enable rest of the types in following PR?

Reply:
I found that the cpu implementation supports Bfloat16 instead of half, so we will not enable the rest of the types in the near future I think.

Reply:
Reason for that, is that x86 CPUs do not support `float16`, but ARM cpus do (aren't they?)
In any case, even if we can not test it, we should have it available, shouldn't we?